plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/product-collection-content.tsx

```diff
@@ -40,6 +43,12 @@ const ProductCollectionContent = (
 		}
 	}, [ queryId, instanceId, setAttributes ] );
 
+	useEffect( () => {
+		if ( ! fastDeepEqual( location, attributes.location ) ) {
+			setAttributes( { location } );
+		}
+	}, [ attributes.location, location, setAttributes ] );
```

Comment:
This looks like a great idea to keep things consistent on both sides (attributes-hook), but at the same time, it seems prone to infinity loops.

To demonstrate, I've added a quick snippet under the `editor.BlockEdit` filter, and I ran:
```js
setAttributes({
	...attributes,
	location: {
		...attributes.location,
		inherit: false,
	},
});
```

This resulted in an infinity loop: ![iuNcfJ.gif](https://github.com/woocommerce/woocommerce/assets/1105590/da6b8d3e-12eb-4a2a-a256-d2dcb88f01bc)

I believe this is a tricky situation stemming from using the custom hook. The loop could be avoided at the cost of losing the 2-way binding. 🤔

Reply:
> To demonstrate, I've added a quick snippet under the editor.BlockEdit filter, and I ran:

Can you provide the full snippet of what you did? I don't think `editor.BlockEdit` is designed to modify attributes at all.

> The loop could be avoided at the cost of losing the 2-way binding. 🤔

Do you have some alternatives in mind? I don't think a hook is a problem here as it's a common pattern of updating `attributes`, especially those that may require side effect (`getEntityRecords`).

Reply:
>  I don't think editor.BlockEdit is designed to modify attributes at all.

Hmm, I can easily grab the `setAttributes` function from the BlockEdit props. It might not be expected and is certainly doable. 

```js
addFilter(
	'editor.BlockEdit',
	'woocommerce-product-recommendations/product-collection/recently-viewed-products',
	handleBlockEdit
);

const handleBlockEdit = (BlockEdit) => {
	return (props) => {
		const { attributes, setAttributes, name } = props;
		const { collection } = attributes;

		if (
			name !== 'woocommerce/product-collection' ||
			collection !==
				'woocommerce-product-recommendations/product-collection/recently-viewed-products'
		) {
			return <BlockEdit {...props} />;
		}

		setAttributes({
			...attributes,
			location: {
				...attributes.location,
				inherit: false,
			},
		});

		// Return something or null.
};
```

Reply:
> Do you have some alternatives in mind?

Based on our previous discussion, the reason for including the location context within the attributes is due to the limitation of PHP in efficiently determining the parent container of a block. This information is essential in identifying whether a block is inside a Single-Product container. However, since these problems arise here, I have been thinking of ways to do this strictly in PHP. 

I don't have a concrete way just yet, but I'd love to discuss them in our next sync!